model-optimizer/extensions/back/ChangeOutputTypeAttributes.py

```diff
@@ -0,0 +1,98 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+
+import numpy as np
+
+from mo.back.replacement import BackReplacementPattern
+from mo.graph.graph import Graph
+from mo.graph.graph import Node
+from mo.middle.passes.convert_data_type import data_type_str_to_np
+from mo.utils.error import Error
+
+operations_with_data_type_attributes = {
+    'Cast': {'attr_name': 'dst_type', 'in_ports_to_check': (0,)},
+    'Range': {'attr_name': 'output_type', 'in_ports_to_check': (0, 1, 2)},
+}
+
+
+class ChangeOutputTypeAttributes(BackReplacementPattern):
+    """
+    For operations that allow specify output type by attribute this transformation
+    changes output type from fp64 to fp32 since not all plugins support fp64 data type.
+    Changes output type from fp32 to fp16 (and ensure that this is possible) when generating IR for fp16.
+    But leave fp32 if node returns shape value even if --data_type=FP16 (look extensions/back/MarkNodesWithShapeValues.py)
+    """
+    enabled = True
+    force_shape_inference = True
+
+    def run_after(self):
+        from extensions.back.MarkNodesWithShapeValues import MarkNodesWithShapeValues
+        return [MarkNodesWithShapeValues]
+
+    def run_before(self):
+        return []
+
+    def find_and_replace_pattern(self, graph: Graph):
+        ir_data_type = data_type_str_to_np(graph.graph['cmd_params'].data_type)
+
+        for op_name, op_map in operations_with_data_type_attributes.items():
+            dst_type = op_map['attr_name']
+            for node in graph.get_op_nodes(op=op_name):
+                node_name = node.soft_get('name', node.id)
+                assert node.has_valid(dst_type)
```

Comment:
Add a node name to the assert error message

Reply:
added